server webui easy config selection

poulphunter opened the pull request:
Add to webui a quick access for custom configurations (including prompt).

![Firefox_Screenshot_2025-02-22T19-19-44 044Z](https://github.com/user-attachments/assets/8de540fb-5326-4aab-b9d7-82ac8702064e)


As asked by @xydac in an old PR, the chat prompts from https://github.com/f/awesome-chatgpt-prompts has been loaded in the config json file.

ngxson commented:
From UX perspective, this will be confused for most users who are not very technical. And from UI perspective, this breaks the UI on mobile and also make it feels messy.

So I think we should keep this as a demo for now. If more users need this, we can rethink jow to make a better UX/UI.

poulphunter commented:
> From UX perspective, this will be confused for most users who are not very technical, as they expect all settings to be accessible from UI, not from a file. And from UI perspective, this breaks the UI on mobile and also make it feels messy.
> 
> So I think we should keep this as a demo for now. If more users need this, we can rethink how to make a better UX/UI.

Now, it's not activated when no json config file is given.
Example file is given but need to be renamed to be used.
From UX perspective, it's now on the settings button and we can manually set it or choose a preset, it's easier to understand.
UI is better on mobile but I haven't fixed legacy Settings dialog that is messy on small devices.

Is also added prompt suggestions.
I've fix the Conversation menu that always show first on mobile.

poulphunter commented:
> > From UX perspective, this will be confused for most users who are not very technical, as they expect all settings to be accessible from UI, not from a file. And from UI perspective, this breaks the UI on mobile and also make it feels messy.
> > So I think we should keep this as a demo for now. If more users need this, we can rethink how to make a better UX/UI.
> 
> Now, it's not activated when no json config file is given. Example file is given but need to be renamed to be used. From UX perspective, it's now on the settings button and we can manually set it or choose a preset, it's easier to understand. UI is better on mobile but I haven't fixed legacy Settings dialog that is messy on small devices.
> 
> Is also added prompt suggestions. I've fix the Conversation menu that always show first on mobile.

![Firefox_Screenshot_2025-02-23T13-10-01 444Z](https://github.com/user-attachments/assets/f0c936d4-6192-4ab4-b6e0-61d0f4feb600)
![Firefox_Screenshot_2025-02-23T13-11-05 675Z](https://github.com/user-attachments/assets/0b217004-2acd-438b-83f8-1caeae8303e5)

poulphunter commented:
Big changes in last commit, translations, presets, suggestions ideas, lot's of UX/UI improvements.
Cleaner code, refactor of parts that may not be in right component etc...
Separate readme for Web UI project.
@ngxson can you review it ?

![mobile_settings](https://github.com/user-attachments/assets/d15be098-a92e-4978-96e1-919cf0c7eba9)
![mobile_python_example_run](https://github.com/user-attachments/assets/3ba45b18-8e26-43b9-82cb-1c6a50e6e692)
![mobile_python_example_ask2](https://github.com/user-attachments/assets/4955e8da-9ae2-44ce-b7d4-74674cf067e8)
![mobile_python_example_ask1](https://github.com/user-attachments/assets/73d2c607-e408-4cf2-b13d-98bb63999111)
![mobile_french_translation_example](https://github.com/user-attachments/assets/a32d6331-c268-4f11-b3d1-2ba72355c05e)
![mobile_conversation_tab](https://github.com/user-attachments/assets/c0c97778-d1a4-4300-bdea-50d2ecfcec52)
![iPad_conversation_Italy](https://github.com/user-attachments/assets/f0d9b3e2-04a2-4402-b9b3-86f5c811fca2)
![desktop_light_theme_presets_suggestions](https://github.com/user-attachments/assets/d37acc9a-2c80-4359-b4a8-fd591af37e17)
![desktop_dark_theme_presets_suggestions](https://github.com/user-attachments/assets/41000875-f930-464d-96d8-d59104c658d2)

ngxson commented on examples/server/webui/src/i18n.tsx:
```diff
@@ -0,0 +1,59 @@
+import Backend from 'i18next-http-backend';
+import LanguageDetector from 'i18next-browser-languagedetector';
+
+import translationCN from './locales/cn/translation.json';
```
I don't think i18n is something we should add right now. I did i18n in various web app product before and the maintenance cost is just too much.

For a small frontend project like this, i18n introduces too many works in the future.

poulphunter replied:
Translation was something that was aked to me.
It's a little weird to ask translation in an app that cannot support the language you ask translation for...
This also make sens as system prompt can be in other languages and we can see "prompt master" user that don't understand well English.
It's heavy for user who begin to use something as embed model with llamafile to have a frontend that ask them for "temperature" with no language friendly explanation.

I have no vision of the future of llama cpp and parameters, if it changes a lots.
I've just in mind that only English can be maintained (with near 0 new cost) as it's the default translation used if language translation was not found.
For costs in mind, I'm not sure that Python interpreter was needed in "small frontend" (I don't consider a ~2.5Mo app as "small frontend" by the way).

ngxson commented on examples/server/webui/src/components/Header.tsx:
```diff
@@ -24,94 +23,138 @@ export default function Header() {
+            className="h-4 w-4"
+          >
+            <path
+              fillRule="evenodd"
```
We've already moved to heroicons, so plain svg like this is not recommended

poulphunter replied:
I don't think that task can be asked here as code was already messy with sometimes icon sometimes svg.
I think refactoring icon can be a separate task (with low priority since it have a very small impact).

By the way, I don't think it's wise to use lot's of libraries (as heroicons) for now, as it adds dependencies (libraries) vs near zero cost because index is compressed. To maintain there are lot's of other tasks to do with higher priority.

ngxson reviewed (commented):
I don't have time to review everything here. Your screenshots look good, but problems are:
- There are too many changes grouped into single PR, I cannot review all of them. Please at least break them down into smaller PRs.
- Most of the introduced functions like i18n or `questionIdeas` bring too much code into the project. Since we're currently aiming for simplicity, I don't think we should add these functions right away. If you have a look on the old webui, it actually has a lot more features compare to this, but with the cost of no one maintaining it.
- The code is not very clean, i.e. still many places use `getElementById` to bypass using proper react ref or component. This will be a lot of headaches when come to managing states.

poulphunter commented:
> I don't have time to review everything here. Your screenshots look good, but problems are:
> 
>     * There are too many changes grouped into single PR, I cannot review all of them. Please at least break them down into smaller PRs.

I've start doing it, but as you asked me to think about UX/UI, I've found lot's of code very messy (modal dialog in mobile was very hard, UX needs lot's of click to open/select/close something, some part of code where sometimes in an other component with no reason, "drawer" was very painful for UX).
It was hard to understand this code, so I've cleaned it in the same time of other functions development, a separate PR would have cost me too much time.
Also, it ask lot's of cost to separate things like "translations" and "presets" as it depends of each other to work well. I've done that in few days, separate PR would have cost few weeks...
> 
>     * Most of the introduced functions like i18n or `questionIdeas` bring too much code into the project. Since we're currently aiming for simplicity, I don't think we should add these functions right away. If you have a look on the old webui, it actually has a lot more features compare to this, but with the cost of no one maintaining it.

I don't know the reason for other dev. to not maintain legacy projects (I've started to update legacy version before see that it has been replaced by a react app). 
Old code was not LLM beginners friendly, it was very hard to understand to peoples I've given "packages" to begin playing with local LLM, but was easier to maintain (adapt) for me as the code was smaller.
I had thinked that if the project change from "small project" of ~250ko to a >2.5Mo react app (that don't fully work on my old browsers) for reasons like not to be a "small project" anymore but a fully usable project (not just a demo), larger community that can help improve it, etc...
I tried to make the application more accessible for non-computer scientists, I don't know if that's the direction of this project.
It have already more complex features like Python interpreter that is not necessary for lot's of peoples and hard to maintain.
> 
>     * The code is not very clean, i.e. still many places use `getElementById` to bypass using proper react ref or component. This will be a lot of headaches when come to managing states.

I've done states management with the help of some coffee ! I think it can be refactored in a more understandable way.
It's explained in the documentation Readme I've added, an it make app really more intuitive.

I've start to use "getElementById" to bypass dropdown UX non-sense of DaisyUI, it's not React and is also lighter, I don't think I will change that as React isn't needed and is heavier for performances, for browsers.
But I may rewrite block visibility using "getElementById" to take full advantages of animations, theme etc...

I give you this improvements but I don't think I will maintain C++ app, so it's up to you to use it or not.
As mentioned above, if it's not in the spirit of the app, I can start a separate project.
I think I will add some other features such as cutting text into chunks for making summaries of large texts.
